From 48c462588b4a9176a524f14b658e765621f434e5 Mon Sep 17 00:00:00 2001
From: Birkir Gudjonsson <birkir.gudjonsson@gmail.com>
Date: Sun, 12 Nov 2017 02:23:22 -0500
Subject: [PATCH] Touchthrough and animation disable flag

---
 ios/RCCLightBox.m | 20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git a/ios/RCCLightBox.m b/ios/RCCLightBox.m
index f630435..662ae78 100755
--- a/ios/RCCLightBox.m
+++ b/ios/RCCLightBox.m
@@ -68,6 +68,7 @@ const NSInteger kLightBoxTag = 0x101010;
         self.reactView.backgroundColor = [UIColor clearColor];
         self.reactView.sizeFlexibility = RCTRootViewSizeFlexibilityWidthAndHeight;
         self.reactView.center = self.center;
+        self.reactView.passThroughTouches = YES;
         [self addSubview:self.reactView];

         [self.reactView.contentView.layer addObserver:self forKeyPath:@"frame" options:0 context:nil];
@@ -78,6 +79,14 @@ const NSInteger kLightBoxTag = 0x101010;
     return self;
 }

+-(id)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
+    id hitView = [super hitTest:point withEvent:event];
+    if (hitView == self) {
+        return nil;
+    } else {
+        return hitView;
+    }
+}

 -(BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch
 {
@@ -217,6 +226,7 @@ const NSInteger kLightBoxTag = 0x101010;

 +(void)showWithParams:(NSDictionary*)params
 {
+    NSDictionary *style = params[@"style"];
     UIViewController *viewController = RCTPresentedViewController();
     if ([viewController.view viewWithTag:kLightBoxTag] != nil)
     {
@@ -226,16 +236,20 @@ const NSInteger kLightBoxTag = 0x101010;
     RCCLightBoxView *lightBox = [[RCCLightBoxView alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height) params:params];
     lightBox.tag = kLightBoxTag;
     [viewController.view addSubview:lightBox];
-    [lightBox showAnimated];
+    if (style[@"animated"] != nil && [RCTConvert BOOL:style[@"animated"]]) {
+      [lightBox showAnimated];
+    }
 }

 +(void)dismiss
 {
     UIViewController *viewController = RCTPresentedViewController();
     RCCLightBoxView *lightBox = [viewController.view viewWithTag:kLightBoxTag];
+    NSDictionary *style = lightBox.params[@"style"];
-    if (lightBox != nil)
-    {
+    if (style[@"animated"] != nil && [RCTConvert BOOL:style[@"animated"]]) {
         [lightBox dismissAnimated];
+    } else {
+        [lightBox removeFromSuperview];
     }
 }

--
2.6.1

